ownCloud/Client/Actions/Actions+Extensions/UploadPhotosAction.swift

```diff
@@ -95,59 +121,82 @@ class UploadPhotosAction: UploadBaseAction {
 
-			let progress = Progress(totalUnitCount: 100)
-			progress.localizedDescription = String(format: "Importing '%@' from photo library".localized, filename)
-
-			let options = PHAssetResourceRequestOptions()
-			options.isNetworkAccessAllowed = true
-			options.progressHandler = { (completed:Double) in
-				progress.completedUnitCount = Int64(completed * 100)
-			}
-
-			self.publish(progress: progress)
-
-			var localURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(filename)
-
-			let contentInputOptions = PHContentEditingInputRequestOptions()
-			contentInputOptions.isNetworkAccessAllowed = true
-			asset.requestContentEditingInput(with: contentInputOptions) { (input, _) in
-				if let fullImage = CIImage(contentsOf: input!.fullSizeImageURL!) {
-					let storeAsHEIF = ressource.uniformTypeIdentifier == "public.heic" && !userDefaults.convertHeic
-					let colorSpace = CGColorSpaceCreateDeviceRGB()
+		// Setup import options, allow download asset from network if necessary
+		let contentInputOptions = PHContentEditingInputRequestOptions()
+		contentInputOptions.isNetworkAccessAllowed = true
+		contentInputOptions.progressHandler = { (percentage:Double, _) in
+			progress.completedUnitCount = Int64(percentage * 100)
+		}
 
-					var imageData : Data?
-					if storeAsHEIF {
-						imageData = CIContext().heifRepresentation(of: fullImage, format: CIFormat.RGBA8, colorSpace: colorSpace)
-						print(fullImage.properties)
-					} else {
-						localURL = localURL.deletingPathExtension().appendingPathExtension("jpg")
-						imageData = CIContext().jpegRepresentation(of: fullImage, colorSpace: colorSpace)
-					}
-					do {
-						self.unpublish(progress: progress)
-						try imageData?.write(to: localURL)
-						self.upload(itemURL: localURL, to: rootItem, name: localURL.lastPathComponent, completionHandler: { (_) in
-							// Delete the temporary asset file
-							try? FileManager.default.removeItem(at: localURL)
-							self.completed()
-						})
-					} catch {
-						self.completed(with: NSError(ocError: .internal))
+		// Import full size asset
+		asset.requestContentEditingInput(with: contentInputOptions) { (input, _) in
+			self.unpublish(progress: progress)
+
+			self.uploadSerialQueue.async {
+				if let input = input, let url = input.fullSizeImageURL {
+					autoreleasepool {
+						var image = CIImage(contentsOf: url)
+						if image != nil {
+							let storeAsHEIF = input.uniformTypeIdentifier == "public.heic" && !userDefaults.convertHeic
+							let colorSpace = CGColorSpaceCreateDeviceRGB()
+
+							let fileName = url.lastPathComponent
+							var localURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(fileName)
+
+							var ciContext = CIContext()
+							var imageData : Data?
+
+							func cleanUpCoreImageRessources() {
+								// Release memory consuming resources
+								imageData = nil
+								image = nil
+								ciContext.clearCaches()
+							}
+
+							if storeAsHEIF {
```

Comment:
Unless I miss something, the current implementation
- will always decode and re-compress the file - even if the file is already in the right format (possibly adding additional compression artifacts, changing the colorspace from grayscale to RGB, increase file size, …) 
- also convert JPEG originals to HEIF files if HEIF is selected as output format

IMO the implementation should
- preserve the original file whenever possible
- implement the option `Convert HEIC to JPEG` as described:
    - when `off`: upload files "as-is", using the original file and format
    - when `on`: convert HEIF files to JPEGs before upload, leave JPEGs unchanged

Reply:
@felix-schwarz Fixed finding except for:

> * also convert JPEG originals to HEIF files if HEIF is selected as output format

Since it was not requested in the original issue.

Reply:
Looks fine now. Thanks for making the changes!